### PR TITLE
Setting up mypy

### LINKED ISSUE #43: Setting up poetry dependancy management and packaging tool
We need to setup `poetry` to manage the package.

- [ ] set up pre commit hooks for `mypy`
- [ ] set up CI job for `mypy`

### PR DESCRIPTION
**Pull request by leikoilja**
The beginning of setting up mypy

- sets up pre commit hooks for mypy
- sets up CI job for mypy

Closes #43

### CODE REVIEW

**Comment by leikoilja** [comment]
@KapJI, it's a bit over my head, can you please help me to understand why `exclude = 'glocaltokens/google/*'` line in mypy config is ignored? 🤔  This results in checking proto files that ends up with tons of errors

**Comment by KapJI** [comment]
Add `__init__.py` in all `google` subdirectories. And then in `mypy.ini` (I actually suggest to put configuration if `setup.cfg`) add:
```
[mypy-glocaltokens.google.*]
ignore_errors = True
```

File: .github/workflows/linting-and-testing.yaml [hunk 1]
```diff
@@ -33,3 +33,7 @@ jobs:
       - name: Run pre-commit on all files
         run: |
           poetry run pre-commit run --all-files --show-diff-on-failure --color=always
+
+      # - name: Run MyPy
+      #   run: |
+      #     poetry run mypy glocaltokens
```
Comment (KapJI) [inline .github/workflows/linting-and-testing.yaml:39]: With `mypy` in `pre-commit` config, you won't need this.

**Comment by leikoilja** [comment]
@KapJI, i m working through fixing `mypy` errors and there is one i don't know how to fix. Can you please help me with that one:
```bash
glocaltokens/scanner.py:135: error: Argument 3 to "ServiceBrowser" has incompatible type "CastListener"; expected "Union[ServiceListener, List[Callable[..., None]]
, None]"
Found 1 error in 1 file (checked 4 source files)
```

**Comment by KapJI** [comment]
Derive `CastListener` from [`zeroconf.ServiceListener`](https://github.com/jstasiak/python-zeroconf/blob/4da1612b728acbcf2ab0c4bee09891c46f387bfb/zeroconf/__init__.py#L1531)

**Comment by KapJI** [comment]
It's actually shows that `CastListener` needs to implement `remove_service`. We had issue for that but I can't find it.

**Comment by DurgNomis-drol** [comment]
#46

**Comment by KapJI** [comment]
Yep, this one :)

**Comment by KapJI** [comment]
Also take a look at [`@overrides` decorator](https://github.com/mkorpela/overrides) to check that you're actually overriding that method.

**Comment by leikoilja** [comment]
Mypy doesn't argue on not having `remove_service`, therefore I'll keep it for the next PR

File: glocaltokens/client.py [hunk 2]
```diff
@@ -200,7 +203,7 @@ def get_master_token(self) -> Optional[str]:
             if "Token" not in res:
                 LOGGER.error("[!] Could not get master token.")
                 LOGGER.debug(f"Request response: {res}")
-                return
+                return ""
```
Comment (KapJI) [inline glocaltokens/client.py:206]:
```suggestion
                return None
```

File: glocaltokens/client.py [hunk 3]
```diff
@@ -223,7 +226,7 @@ def get_access_token(self) -> Optional[str]:
             if "Auth" not in res:
                 LOGGER.error("[!] Could not get access token.")
                 LOGGER.debug(f"Request response: {res}")
-                return
+                return ""
```
Comment (KapJI) [inline glocaltokens/client.py:229]:
```suggestion
                return None
```

File: glocaltokens/client.py [hunk 4]
```diff
@@ -313,7 +316,7 @@ def find_device(name) -> Optional[GoogleDevice]:
                     return device
             return None
 
-        devices: [Device] = []
+        devices: List[Device] = []
```
Comment (KapJI) [inline glocaltokens/client.py:319]: Good catch 😀

File: mypy.ini [hunk 5]
```diff
@@ -0,0 +1,5 @@
+[mypy]
+ignore_missing_imports = True
```
Comment (KapJI) [inline mypy.ini:2]: Let's only set this for packages which doesn't provide typing info, e.g. `faker`. Like in `ha-google-home`.

Comment (KapJI) [reply mypy.ini:2]: Worth setting `python_version = 3.8` here.

File: mypy.ini [hunk 6]
```diff
@@ -0,0 +1,5 @@
+[mypy]
```
Comment (KapJI) [inline mypy.ini:1]: Let's move this to `setup.cfg`. The idea is to have less configuration files if we add more tools. All python tools support configuration either in `setup.cfg` or `pyproject.toml` (which should be preferred but not supported by all tools yet).

**Comment by leikoilja** [comment]
Check it out, @KapJI

**Review by KapJI — APPROVED** [review]
Awesome! Ship it 🚀

**Pull request merged**